### PR TITLE
Rearrange documentation and add "learn the workflow" section

### PR DESCRIPTION
**Pull request by Lotes**
I reorganized the documentation in this PR. I do not want to delete any existing pages, but sometimes they are a bit long and misleading. Most of them are fine for me.
I also added a learning section where (*new*) the basic workflow with Langium gets explained.

* everything from "old/Guides" you can find under "new/Recipes"
* everything from "old/Tutorials" you can find under "new/Learn/Minilogo"
* everything from "old/Documentation" got intermixed (WIP)
  * "old/Documentation/Langium Overview" you can find under "new/What is Langium/Features"
  * "old/Documentation/Getting started" you can find under "new/Reference/Glossary" and several "new/Learn/Workflow/*"
  * the remaining part of "old/Documentation" you can find under "new/Reference"

### CODE REVIEW

File: hugo/content/docs/learn/workflow/_index.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+{{<mermaid>}}
+flowchart TD
+  A(["1. Install Yeoman generator"]);
+  B(["2. Scaffold a Langium project"]);
+  C(["3. Write the grammar"]);
+  D(["4. Generate AST files"]);
+  E(["5. Resolve cross-references"]);
+  F(["6. Create validations"]);
+  G(["7. Generate what you want"]);
```
Comment (Lotes) [inline hugo/content/docs/learn/workflow/_index.md:17]: Each sub step should have a way of validation as far as possible. Like a hint of how to write a test for it.

**Comment by JohannesMeierSE** [comment]
@Lotes I wrote a tutorial about keywords as identifiers in https://github.com/eclipse-langium/langium-website/pull/219, but it is not yet merged. Maybe it belongs to "Recipes"? It could be seen as a tutorial as well, but it is not related to Minilogo, therefore "Minilogo tutorial" would not fit as category.

**Comment by Lotes** [comment]
@JohannesMeierSE

I would put it under recipes, since it is not really learning the workflow, it is some kind of feature you want to talk about.

The Minilogo tutorial is also some kind of learning + multiple recipes, and I think maybe we should adapt it to fit more with the workflow. On the other hand... each project has its own challenges and therefore differs from the core workflow.

**Comment by JohannesMeierSE** [comment]
I got some ideas in detail:
- I would place "Code Bundling" as last point in "Recipes", since it is the final thing you would do. Having "Scoping" as one of the most important things at position 1 makes sense for me.
- At the moment, "Documentation" is an empty page: Maybe, we could place here some tips, when to read which parts of the documentation?

Regarding "Our Workflow":
- I would rename it to "Langium's Workflow", since I don't like "Our", that sounds a bit like suggestions of TypeFox.
- Adding numbers 1, 2, 3, ... before the names in the outline/TOC in the left side bar would improve the readability and the steps.
- Having the names in the workflow diagram and in the side bar 100% the same makes it easier to match the steps.

For the "Workflow diagram":
- I would treat "Find advanced topics" not as a step, in particular, not as the picture suggests: After applying an advanced topic, you cannot iterate your DSL anymore (that is, what the diagram tells the reader). You want to say, that there is much more, which does not fit into these seven steps. Therefore, I suggest to say, that these 7 steps reflect the base workflow, in particular, for becoming familiar with Langium, but there are much more advanced topics in "Recipes".
- Maybe, you could have a dedicated section for the "Initial Set-up" before the section for the "Core workflow"? Than all steps are visually represented in the same way.

Nevertheless, I like your clear steps a lot and I think, they will help newcomers a lot!
My comments target only some details of the updated documentation structure.

**Comment by Lotes** [comment]
Thank you, @JohannesMeierSE . I made all these changes.

**Comment by Lotes** [comment]
Add a hint on names for resolving scope. The name is the key to resolve cross-references.

File: hugo/content/docs/introduction/_index.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+If you need a more detailed list of Langium features, you can find them in the [features section](/docs/features).
+
+### Try it out
+
+If you want to see Langium in action, you can follow the [showcases](/showcase) or event the [playground](/playground).
+
+### Learn Langium
+
+If you are convinced with Langium and want to learn more about it, you can start with the [learn section](/docs/learn).
```
Comment (gfontorbe) [inline hugo/content/docs/introduction/_index.md:19]:
```suggestion
If you are convinced by Langium and want to learn more about it, you can start with the [learn section](/docs/learn).
```

File: hugo/content/docs/recipes/_index.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+title: "Recipes"
+weight: 400
+url: "/docs/recipes"
+---
+
+## Where to go from here?
+
+Take your time to study the recipes within the navigation on the left. They are designed to help you with common tasks and challenges you might face when working with Langium. If you have any questions or suggestions, feel free to [create an issue](https://github.com/eclipse-langium/langium/issues) or [start a discussion](https://github.com/eclipse-langium/langium/discussions).
```
Comment (gfontorbe) [inline hugo/content/docs/recipes/_index.md:9]:
```suggestion
Take your time to study the recipes within the navigation on the left. They are designed to help you with common tasks and challenges you might face when working with Langium. If you have any questions or suggestions, feel free to [create an issue](https://github.com/eclipse-langium/langium/issues) or [start a discussion](https://github.com/eclipse-langium/langium/discussions) on the Github repository.
```

File: hugo/content/docs/reference/grammar-language.md [hunk 4]
```diff
@@ -1,5 +1,5 @@
 ---
-title: "The Grammar Language"
+title: "Grammar Language"
 weight: 100
 ---
 The grammar language describes the syntax and structure of your language. The [Langium grammar language](https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/grammar/langium-grammar.langium) is implemented using Langium itself and therefore follows the same syntactic rules as any language created with Langium. The grammar language will define the structure of the *abstract syntax tree* (AST) which in Langium is a collection of *TypeScript types* describing the content of a parsed document and organized hierarchically. The individual nodes of the tree are then represented with JavaScript objects at runtime.
```
Comment (gfontorbe) [inline hugo/content/docs/reference/grammar-language.md:5]: Maybe it would be good to add a table of content at the top of the page?

**Review by gfontorbe — APPROVED** [review]
I really like the new flow of the documentation! Thanks a lot @Lotes.

Just a couple minor suggestions

File: hugo/content/docs/learn/workflow/generate_ast.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+* a _grammar_ file: which contains your entire grammar definition in JSON format.
+* a _module_ file: which contains language-specific setup objects for the final module definition of your language.
+* an _ast_ file: which contains the definition of your AST.
+* several _syntax highlighting_ files: like for PrismJS, TextMate or Monarch.
+
+## The syntax tree
+
+An AST of your language is now ready to be get parsed. One important concept in Langium are _cross-references_. With them you can reference other elements in your language. For example, you can reference a variable in a function call. The AST will contain a reference to the variable. This is useful for code analysis and transformation. Technologies like ANTLR or other parser-only generators do not support this feature. For them you are forced to resolve these references in-place everytime the developer is confronted with them.
+
+After this generation steps, cross-references are not resolved yet. This is done in the next step.
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/generate_ast.md:25]:
```suggestion
After these generation steps, cross-references are not resolved yet. This is done in the next step.
```

File: hugo/content/docs/learn/workflow/generate_ast.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    Hello Jane!
+`);
+
+//assert
+const model = document.parseResult.value;
+expect(model.persons).toHaveLength(2);
+expect(model.persons[0].name).toBe("John");
+expect(model.persons[1].name).toBe("Jane");
+expect(model.greetings).toHaveLength(2);
+//be aware of the fact that the following checks will fail, because the cross-references are not resolved yet
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/generate_ast.md:98]:
```suggestion
//be aware of the fact that the following checks will fail at this point, because the cross-references are not resolved yet
```

File: hugo/content/docs/learn/workflow/generate_everything.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
+---
+title: "7. Generate your artifacts"
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/generate_everything.md:2]:
```suggestion
title: "7. Generate the artifacts"
```

Comment (insafuhrmann) [reply hugo/content/docs/learn/workflow/generate_everything.md:2]: Or even "Generate artifacts". Reason: You have not used "your" in the other captions and it looks a bit long.

File: hugo/content/docs/learn/workflow/resolve_cross_references.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+  G2 --> KW2('hello')
+  G2 --> PRef2[Ref]
+  PRef2 -- $refText --> RT2('Jane')
+  G2 --> EM2('!')
+  PRef2 -..-> P2
+{{</mermaid>}}
+
+## Resolution of cross-references
+
+As already hinted, you can implement a scope provider and a scope computation. Fortunately, Langium comes with default implementations for both. But eventually as your language grows, you might want to implement your own strategy because the default is not sufficient. In the follwoing sections I will sketch how to interpret their interfaces.
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/resolve_cross_references.md:87]: Maybe reword this to remove the "I" from this? (Who is "I" for the documentation of an open source project?)

File: hugo/content/docs/learn/workflow/resolve_cross_references.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+#### Terms
+
+The _scope provider_ is responsible for providing a _scope_ for a given cross-reference represented by the `ReferenceInfo` type.
+
+A _scope_ is a collection of AST nodes that are represented by the `AstNodeDescription` type.
+
+The _description_ is like a (string) path through the AST of a document. It can be also seen as a tuple of document URI, JSON path, name and type of the AST node.
+
+A _reference info_ contains the concrete AST reference (which points to nothing yet). The info also has a the parent AST node (a so-called container) of the reference and the property name under which you can find the reference under its container. In the form of this tuple (`container`, `property`, `reference`) Langium visits all cross-references using the scope providers `getScope` method.
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/resolve_cross_references.md:99]:
```suggestion
A _reference info_ contains the concrete AST reference (which points to nothing yet). The info also has a the parent AST node (a so-called container) of the reference and the property name under which you can find the reference under its container. In the form of this tuple (`container`, `property`, `reference`) Langium visits all cross-references using the scope provider's `getScope` method.
```

File: hugo/content/docs/learn/workflow/resolve_cross_references.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+export interface Scope {
+    getElement(name: string): AstNodeDescription | undefined;
+    getAllElements(): Stream<AstNodeDescription>;
+}
+```
+
+#### Purpose
+
+So, what is the purpose of the scope provider? As mentioned above: it visits each cross-reference and tries to find the corresponding AST nodes over the entire workspace, that can be a candidate for the cross-reference's place. It is important to understand that we do not decide here which of these nodes is the perfect match! That decision is part of the so-called linker of the Langium architecture.
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/resolve_cross_references.md:121]:
```suggestion
So, what is the purpose of the scope provider? As mentioned above: it visits each cross-reference and tries to find the corresponding AST nodes over the entire workspace that can be a candidate for the cross-reference's place. It is important to understand that we do not decide here which of these nodes is the perfect match! That decision is part of the so-called linker of the Langium architecture.
```

File: hugo/content/docs/learn/workflow/resolve_cross_references.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+If your cross-reference's `$refText` contains the name `Jane` does not matter here. We need to provide all nodes that are possible at this position. So in the result, you would return `Jane` and `John` AST nodes - for both cross-references!
+
+The background for this behavior is, that this mechanism can be used for two things: the cross-reference resolution and the code completion. The code completion needs to know all possible candidates for a given cross-reference. The resolution of the cross-reference is done by the linker: Given a scope for a certain cross-reference, the linker decides which of the candidates is the right one - for example the first candidate with the same name.
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/resolve_cross_references.md:125]:
```suggestion
The background for this behavior is that this mechanism can be used for two things: the cross-reference resolution and the code completion. The code completion needs to know all possible candidates for a given cross-reference. The resolution of the cross-reference is done by the linker: Given a scope for a certain cross-reference, the linker decides which of the candidates is the right one - for example the first candidate with the same name.
```

File: hugo/content/docs/learn/workflow/write_grammar.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+```
+
+Before we tell Langium anything about our grammar contents, we first need to give it a name - in this case it's `HelloWorld`. The `langium-cli` will pick this up to prefix any generated services with this name.
+
+```langium
+hidden terminal WS: /\s+/;
+terminal ID: /[_a-zA-Z][\w]*/;
+```
+
+Here we define our two needed terminals for this grammar: The whitespace `WS` and identifier `ID` terminals. Terminals parse a part of our document by matching it against their regular expression. The `WS` terminal parses any whitespace characters with the regex `/\s+/`. This allows us consume whitespaces in our document. As the terminal is declared as `hidden`, the parser will parse any whitespace and discard the results. That way, we don't have to care about how many whitespaces a user uses in their document. Secondly, we define our `ID` terminal. It parses any string that starts with an underscore or letter and continues with any amount of characters that match the `\w` regex token. It will match `Alice`, `_alice`, or `_al1c3` but not `4lice` or `#alice`. Langium is using the JS regex dialect for terminal definitions.
```
Comment (insafuhrmann) [inline hugo/content/docs/learn/workflow/write_grammar.md:38]:
```suggestion
Here we define our two needed terminals for this grammar: The whitespace `WS` and identifier `ID` terminals. Terminals parse a part of our document by matching it against their regular expression. The `WS` terminal parses any whitespace characters with the regex `/\s+/`. This allows us to consume whitespaces in our document. As the terminal is declared as `hidden`, the parser will parse any whitespace and discard the results. That way, we don't have to care about how many whitespaces a user uses in their document. Secondly, we define our `ID` terminal. It parses any string that starts with an underscore or letter and continues with any amount of characters that match the `\w` regex token. It will match `Alice`, `_alice`, or `_al1c3` but not `4lice` or `#alice`. Langium is using the JS regex dialect for terminal definitions.
```

File: hugo/content/docs/recipes/multiple-languages.md [hunk 13]
```diff
@@ -253,10 +260,10 @@ After this step, Langium is set up correctly. But if you try to build now, the c
 Let's clean up the error lines. Here are some general hints:
 
 * keep in mind, that you are dealing with three file types now, namely `*.me`, `*.who` and `*.hello`
-    * you can distinguish them very easily by selecting the right sub service from the result object of `createMultipleLanguagesServices`, which is either `Configuration`, `Definition` or `Implementation`, but not `shared`
-    * all these services have a sub service with file extensions: `[Configuration,Definition,...].LanguageMetaData.fileExtensions: string[]`
-    * so, when you are obtaining any documents from the `DocumentBuilder` you can be sure that they are parsed by the matching language service
-    * to distinguish them on your own, use the AST functions for determining the root type, for example for the Configuration language use `isConfigurationUnit(document.parseResult.value)`
+  * you can distinguish them very easily by selecting the right sub service from the result object of `createMultipleLanguagesServices`, which is either `Configuration`, `Definition` or `Implementation`, but not `shared`
```
Comment (insafuhrmann) [inline hugo/content/docs/recipes/multiple-languages.md:263]:
```suggestion
  * you can distinguish them very easily by selecting the right sub service from the result object of `createMultipleLanguagesServices`, which is either `Configuration`, `Definition`, or `Implementation`, but not `shared`
```

File: hugo/content/docs/reference/glossary.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+_document_: An abstract term to refer to a text file on your file system or an open editor document in your IDE.
+
+_grammar_: Defines the form of your language. In Langium, a grammar is also responsible for describing how the _AST_ is built.
+
+_parser_: A program that takes a _document_ as its input and computes an _abstract syntax tree_ as its output.
+
+_parser rule_: A parser rule describes how a certain _AST_ element is supposed to be parsed. This is done by invoking other _parser rules_ or _terminals_.
+
+_terminal_: A terminal is the smallest parseable part of a document. It usually represents small pieces of text like names, numbers, keywords or comments.
```
Comment (insafuhrmann) [inline hugo/content/docs/reference/glossary.md:18]:
```suggestion
_terminal_: A terminal is the smallest parsable part of a document. It usually represents small pieces of text like names, numbers, keywords or comments.
```

**Review by insafuhrmann — APPROVED** [review]
Thank you very much for your PR, @Lotes. I like the new flow very much and also the additions you made. I just added a few detail remarks.

File: hugo/content/docs/learn/workflow/_index.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+title: "Langium's workflow"
+weight: 0
+url: /docs/learn/worflow
+---
+
+Langium's workflow can be expressed as a flow chart diagram, which boils down to the following steps in the diagram.
+Be aware of the fact that the possibilities go beyond this simple workflow. For more advanced topics, you can find answers in the [recipes](/docs/recipes).
+
+{{<mermaid>}}
+flowchart TD
```
[the remark this answers is not in the document]
Comment (Lotes) [reply hugo/content/docs/learn/workflow/_index.md:11]: I can additionally name it "grammar change iteration"

File: hugo/config.toml [hunk 16]
```diff
@@ -80,4 +80,7 @@ enableRobotsTXT = true
   # (Optional, default 'title') Configure how to sort file-tree menu entries. Possible options are 'title', 'linktitle',
   # 'date', 'publishdate', 'expirydate' or 'lastmod'. Every option can be used with a reverse modifier as well
   # e.g. 'title_reverse'.
-  #geekdocFileTreeSortBy = "title"
\ No newline at end of file
+  #geekdocFileTreeSortBy = "title"
+
+  geekdocCollapseSection = true
```
[the remark this answers is not in the document]
Comment (insafuhrmann) [reply hugo/config.toml:85]: As already stated, you can leave it as it is if other solutions do not prove to look nicer. There are arguments in favour of it (esp. point to relevant caption).

File: hugo/content/docs/learn/workflow/_index.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (insafuhrmann) [reply hugo/content/docs/learn/workflow/_index.md:11]: Yes, something like that would be helpful.
